### PR TITLE
grpc-web: suppress content-length header from grpc-web requests

### PR DESCRIPTION
**Pull request by negator**
grpc-web: suppress content-length header from web requests

*Description*:
When using envoy in a tiered deployment, with a grpc-web filter on the downstream envoy, requests to upstream envoys fail over http2 when a content-length header is present, since the underlying nghttp2 library strictly adheres to the [http2 spec](https://http2.github.io/http2-spec/#malformed). This header is usually passed up from browsers, but reflects the total http1.1 payload size, not the total DATA frame sizes.

*Risk Level*: Low

*Testing*:
A simple unit test was added to verify that content-length headers are suppressed.

*Release Notes*:
N/A

Fixes https://github.com/envoyproxy/envoy/issues/2942

### CODE REVIEW

**Comment by alyssawilk** [comment]
Curious, can you explain why this is a grpc specific issue and not an H2 issue in general?  I just want to make sure we're not doing too narrow of a fix.
@mattklein123 who may know offhand :-P

**Comment by negator** [comment]
You’re right, it’s not a grpc specific issue. It could occur anytime an incorrect content-length header is passed to a secondary envoy backend over h2. However this fix is necessary regardless of any other general fix, since browsers (currently) know nothing about h2 data framing and the content-length header is almost always going to be incorrect.

**Comment by negator** [comment]
I don’t think removing it in all cases is the right thing to do, since it could lead to surprising behavior for clients that expect it to be there. But I do think we’d want to perform validation in the h2 codec, or elsewhere, prior to invoking an h2 backend. But I’d defer to you.

File: source/common/grpc/grpc_web_filter.cc [hunk 1]
```diff
@@ -46,6 +46,7 @@ Http::FilterHeadersStatus GrpcWebFilter::decodeHeaders(Http::HeaderMap& headers,
   }
   is_grpc_web_request_ = true;
 
+  headers.removeContentLength();
```
Comment (mattklein123) [inline source/common/grpc/grpc_web_filter.cc]: Please add a comment as to why this is being removed.

**Comment by alyssawilk** [comment]
Ah see I misunderstood the problem - I thought upstream was choking on content length being present not because the gRPC stack was changing the payload and not updating the header.  In that case yeah, it should be a fix to the filter and we can add codec-level validation later on.

File: source/common/grpc/grpc_web_filter.cc (hunk 1 above)
Comment (sayrer) [reply source/common/grpc/grpc_web_filter.cc]: Switching to chunked transfer encoding is what nginx does in this situation, fwiw

Comment (lizan) [reply source/common/grpc/grpc_web_filter.cc]: @sayrer in grpc-web context the upstream must be http2, there is no chunked encoding, hence removing content length is enough.

**Comment by lizan** [comment]
> We could consider somehow checking/warning/enforcing this at the Envoy level so at least you would know at the local level if you were doing something wrong...

or at nghttp2 level, since it does enforce (reject with PROTOCOL_ERROR) this for incoming request, so it can raise at least a warning for outgoing request too.

**Pull request merged**